commit ef67b3d20f49f2fff860e6a251c8655f5ba568ba (refs/changes/10/710/2)
Author: David Pacheco <dap@joyent.com>
Date:   2016-11-07T11:55:44-08:00 (2 years, 11 months ago)
    
    MANTA-3006 marlin should adopt moray v2
    MANTA-3007 marlin test "jobMcanceledRequests" hangs or fails

diff --git a/common/lib/bus.js b/common/lib/bus.js
index d4fc16c..843fe16 100644
--- a/common/lib/bus.js
+++ b/common/lib/bus.js
@@ -277,9 +277,9 @@ MorayBus.prototype.connect = function ()
 	    'host': this.mb_host,
 	    'port': this.mb_port,
 	    'log': this.mb_log.child({ 'component': 'moray' }),
-	    'reconnect': true,
 	    'retry': this.mb_reconnect,
-	    'dns': this.mb_dns
+	    'dns': this.mb_dns,
+	    'unwrapErrors': true
 	});
 
 	client.on('error', function (err) {
diff --git a/common/lib/marlin.js b/common/lib/marlin.js
index bd2b07d..ffff647 100644
--- a/common/lib/marlin.js
+++ b/common/lib/marlin.js
@@ -227,8 +227,8 @@ function MarlinApi(args)
 	    'host': url['hostname'],
 	    'port': parseInt(url['port'], 10),
 	    'log': args['log'].child({'component': 'moray'}),
-	    'reconnect': true,
-	    'retry': conf['moray']['retry']
+	    'retry': conf['moray']['retry'],
+	    'unwrapErrors': true
 	});
 
 	this.ma_client.on('close', this.emit.bind(this, 'close'));
diff --git a/common/lib/util.js b/common/lib/util.js
index 790392e..3a8419f 100644
--- a/common/lib/util.js
+++ b/common/lib/util.js
@@ -93,7 +93,8 @@ function maHttpProxy(args, callback)
 	    'summary': request.method + ' ' + request.url,
 	    'error': false,
 	    'wroteHeader': false,
-	    'request_args': server_args
+	    'request_args': server_args,
+	    'callback_invoked': false
 	};
 
 	server_args['method'] = request.method;
@@ -121,6 +122,8 @@ function maHttpProxy(args, callback)
 			return;
 
 		if (!err) {
+			mod_assert.ok(!state['callback_invoked']);
+			state['callback_invoked'] = true;
 			callback(null, res);
 			return;
 		}
@@ -131,9 +134,15 @@ function maHttpProxy(args, callback)
 
 		if (!state['wroteHeader']) {
 			response.send(err);
+			mod_assert.ok(!state['callback_invoked']);
+			state['callback_invoked'] = true;
 			callback(err);
 		} else {
 			response.end();
+			if (!state['callback_invoked']) {
+				state['callback_invoked'] = true;
+				callback();
+			}
 		}
 	};
 
@@ -146,6 +155,14 @@ function maHttpProxy(args, callback)
 	response.on('close', function () {
 		request.log.warn('response closed unexpectedly');
 		subrequest.abort();
+
+		/*
+		 * Depending on the state we're in when the abort is processed,
+		 * we may receive no more events, so we need to process the
+		 * completion now.
+		 */
+		checkDone('response close',
+		    new Error('response closed unexpectedly'));
 	});
 
 	if (server_args['headers']['expect'] == '100-continue') {
diff --git a/dev/package.json b/dev/package.json
index 77f8469..a15ad63 100644
--- a/dev/package.json
+++ b/dev/package.json
@@ -32,7 +32,7 @@
 		"manta-compute-bin": "git+https://github.com/joyent/manta-compute-bin.git#master",
 		"memorystream": "0.2.0",
 		"mkdirp": "0.3.1",
-		"moray": "git+https://github.com/joyent/node-moray.git#fd5781bc25a9bfe2ba82167664639753fb9f0ca5",
+		"moray": "^2.0.0",
 		"native-dns": "0.2.1",
 		"node-uuid": "1.4.0",
 		"panic": "0.2.1",
diff --git a/dev/tools/test-proxy-proxy.js b/dev/tools/test-proxy-proxy.js
index a56b513..b400c67 100644
--- a/dev/tools/test-proxy-proxy.js
+++ b/dev/tools/test-proxy-proxy.js
@@ -20,11 +20,20 @@ var restify = require('restify');
 var mod_localutil = require('../../common/lib/util');
 
 var log;
+var noutstanding = 0;
+var baseport = 9100;
 
 function main()
 {
 	var server;
 
+	if (process.argv.length > 2) {
+		baseport = parseInt(process.argv[2], 10);
+		if (isNaN(baseport)) {
+			throw (new Error('bad base port'));
+		}
+	}
+
 	log = new bunyan({ 'name': 'testproxy', 'level': 'trace' });
 	server = restify.createServer({
 	    'name': 'testproxy',
@@ -41,7 +50,7 @@ function main()
 	/* JSSTYLED */
 	server.get(/.*/, proxy);
 
-	server.listen(8080, function () {
+	server.listen(baseport + 1, function () {
 	    console.log('%s listening at %s', server.name, server.url);
 	});
 }
@@ -55,13 +64,16 @@ function proxy(request, response, next)
 	    'response': response,
 	    'server': {
 		'host': '127.0.0.1',
-		'port': 8087
+		'port': baseport
 	    }
 	};
 
 	log.info('maHttpProxy dispatched');
+	noutstanding++;
 	mod_localutil.maHttpProxy(proxyargs, function (err, res) {
-		log.info('maHttpProxy callback called');
+		noutstanding--;
+		log.info({ 'noutstanding': noutstanding },
+		    'maHttpProxy callback called');
 		next();
 	});
 }
diff --git a/dev/tools/test-proxy-upstream.js b/dev/tools/test-proxy-upstream.js
index 03c5ec3..7397ae5 100644
--- a/dev/tools/test-proxy-upstream.js
+++ b/dev/tools/test-proxy-upstream.js
@@ -23,5 +23,5 @@ for (i = 0; i < 10 * 1024 * 1024; i++)
 http.createServer(function (req, res) {
 	res.writeHead(200, { 'content-type': 'text/plain' });
 	res.end(data);
-}).listen(8087, '127.0.0.1');
-console.log('server running at 127.0.0.1:8087');
+}).listen(9100, '127.0.0.1');
+console.log('server running at 127.0.0.1:9100');
diff --git a/package.json b/package.json
index 84bdc51..88e7c0f 100644
--- a/package.json
+++ b/package.json
@@ -22,7 +22,7 @@
 		"jsprim": "0.5.0",
 		"lstream": "0.0.3",
 		"mahi": "git+https://github.com/joyent/node-mahi.git#master",
-		"moray": "git+https://github.com/joyent/node-moray.git#fd5781bc25a9bfe2ba82167664639753fb9f0ca5",
+		"moray": "^2.0.0",
 		"node-uuid": "1.4.0",
 		"posix-getopt": "1.0.0",
 		"retry": "0.6.0",
